### PR TITLE
neonvm: Use container statuses, not pod phase, to trigger restart

### PR DESCRIPTION
**Pull request by sharnoff**
Extracted from #738, which adds a second container to the runner pods. Because of that second container, if only one container exits, the pod will still have `.status.phase = Running`, so we need to proactively notice that one of the containers has stopped and propagate that status to the VM itself.

This also introduces some funky logic around how we handle restarts: Because the `Succeeded` and `Failed` phases no longer imply that QMEU itself has stopped, we need to explicitly wait until either the pod is gone or the neonvm-runner container has stopped; otherwise we could end up with >1 instance of the VM running at a time.

---

Tested locally as part of #738 and validated that it has the intended effect and works as expected when I `kubectl exec <pod> -c <container> -- kill 1`.

NB: Some follow-up changes in #754 building on this. See there for details.

### CODE REVIEW

File: neonvm/controllers/virtualmachine_controller.go [hunk 1]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
+		return runnerUnknown
+
+	// With runners with neonvm-container-mgr, the pod phase itself won't be updated if just the
+	// neonvm-runner container exits or completes.
+	// So we need to manually check through the container statuses to check whether they're ok.
+	case corev1.PodRunning:
+		completed := false
+
+		for _, stat := range pod.Status.ContainerStatuses {
+			if stat.State.Terminated != nil {
```
Comment (Omrigan) [inline neonvm/controllers/virtualmachine_controller.go:818]: Wait, this checks all containers, while from the function name I'd expect it to check for `runner` state. Which one is correct?

File: neonvm/controllers/virtualmachine_controller.go [hunk 2]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
 		}
 
 	default:
-		// do nothing
+		panic(fmt.Errorf("unknown pod phase: %q", pod.Status.Phase))
 	}
+}
 
-	return nil
+func runnerContainerStopped(pod *corev1.Pod) bool {
```
Comment (Omrigan) [inline neonvm/controllers/virtualmachine_controller.go:839]: from the function name, i'd expect it to be something like `runnerContainerStatus(pod) != runnerRunning`

File: neonvm/controllers/virtualmachine_controller.go [hunk 3]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
-			// get runner to delete
-			vmRunner := &corev1.Pod{}
-			err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
-			if err == nil {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
-				}
-			} else if !apierrors.IsNotFound(err) {
+		// Always delete runner pod. Otherwise, we could end up with one container succeeded/failed
```
[the remark this answers is not in the document]
Comment (Omrigan) [reply neonvm/controllers/virtualmachine_controller.go:747]: Perhaps somewhere exists a state diagram which would explain?

File: neonvm/controllers/virtualmachine_controller.go [hunk 4]
```diff
@@ -609,24 +609,24 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
 		}
 
 		// runner pod found, check that it's still up:
-		switch vmRunner.Status.Phase {
-		case corev1.PodSucceeded:
+		switch runnerContainerStatus(vmRunner) {
```
Comment (Omrigan) [inline neonvm/controllers/virtualmachine_controller.go:612]: This function is in dire need of refactoring. 500+ lines and nested switch cases make it hard to follow, and e.g. to understand which `case`/`default` applies to which `switch`.

Comment (shayanh) [reply neonvm/controllers/virtualmachine_controller.go:612]: see #733

File: neonvm/controllers/virtualmachine_controller.go [hunk 5]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
+
+const (
+	runnerUnknown   runnerStatus = "Unknown"
+	runnerPending   runnerStatus = "Pending"
+	runnerRunning   runnerStatus = "Running"
+	runnerFailed    runnerStatus = "Failed"
+	runnerSucceeded runnerStatus = "Succeeded"
+)
+
+func runnerContainerStatus(pod *corev1.Pod) runnerStatus {
```
Comment (shayanh) [inline neonvm/controllers/virtualmachine_controller.go:800]: Is there a way to declare in k8s if one of the containers in the pod fails, then the entire pod also should fail?

File: neonvm/controllers/virtualmachine_controller.go [hunk 6]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
-			if err == nil {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
-				}
-			} else if !apierrors.IsNotFound(err) {
+		// Always delete runner pod. Otherwise, we could end up with one container succeeded/failed
+		// but the other one still running (meaning that the pod still ends up Running).
+		vmRunner := &corev1.Pod{}
+		err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
```
Comment (shayanh) [inline neonvm/controllers/virtualmachine_controller.go:750]: When dealing with state machines like this, it's easier to track down what's going on where each case is handled separately. Now it's kinda hard to tell what we do in case of `vmv1.RestartPolicyAlways` and `vmv1.RestartPolicyAlways`. I suggest handling each case separately as before.

File: neonvm/controllers/virtualmachine_controller.go (hunk 5 above)
Comment (shayanh) [reply neonvm/controllers/virtualmachine_controller.go:800]: Also this function doesn't return runner container status exactly, which is misleading.

File: neonvm/controllers/virtualmachine_controller.go [hunk 7]
```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
...
 			}
-			// delete runner only when VM failed
-			if found && virtualmachine.Status.Phase == vmv1.VmFailed {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
+
+			// TODO for RestartPolicyNever: implement TTL or do nothing
+		}
+	default:
```
[the remark this answers is not in the document]
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:783]: Honestly not sure; don't want to remove it in case it's load-bearing :sweat_smile:
Probably we'll drop it as part of refactoring all this.

File: neonvm/controllers/virtualmachine_controller.go (hunk 5 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:800]: > Is there a way

I spent a while looking and couldn't find any way to do this. K8s 1.29 introduced ["sidecar containers"](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) as an explicit feature, which _do_ allow this.

File: neonvm/controllers/virtualmachine_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:747]: Yeah; it's a bit confusing - I'm not sure there's a better way to do it though.

I also don't think there's an existing state diagram. I'll try to give an explanation here, for this part of the state diagram. Hopefully we can make this clearer both in code and in documentation as part of #733.

---

## Before this PR:

| Runner pod status | VM status |
|---|---|
| Overlay IP not acquired _or_ mid-restart | `""` |
| Pod doesn't exist | `"Pending"` |
| `"Pending"` | `"Pending"` |
| `"Running"` | `"Running"`, `"Scaling"`, `"PreMigrating"`, `"Migrating"` |
| `"Succeeded"` | `"Succeeded"` |
| `"Failed"` | `"Failed"` |

Note that the _Pod_ is "Succeeded" if all containers exit successfully, and "Failed" if all containers exit and at least one failed. This means that if there is at least one container still running (like `container-mgr`), the pod will continue to be in the "Running" phase.

Also note: When we say "Pod doesn't exist", technically the controller (before this PR) does not wait for the pod to be fully deleted when restarting. By the time the VM gets to "Succeeded" or "Failed" (prompting restart), we already know that there's nothing running in the pod, so we can just _start_ deleting the pod and don't need to wait for it to finish:

https://github.com/neondatabase/autoscaling/blob/a7943db25bf5d97eaf4771b46e10a3d97ed4b3ea/neonvm/controllers/virtualmachine_controller.go#L748-L764

(nb: `virtualmachine.Cleanup()` sets `.Status.PodName = ""`, among other things)

## After this PR:

| Runner pod status | VM status |
|---|---|
| Overlay IP not acquired | `""` |
| Pod doesn't exist | `"Pending"` (or `"Succeeded"` / `"Failed"` on exit, if no restart) |
| `"Pending"` | `"Pending"` |
| `"Running"` (and no containers exited) | `"Running"`, `"Scaling"`, `"PreMigrating"`, `"Migrating"` |
| `"Running"` (neonvm-runner exited successfully) | `"Succeeded"` |
| `"Running"` (other container exited successfully) | `"Failed"` |
| `"Running"` (≥1 container failed) | `"Failed"` |
| `"Failed"` | `"Failed"` (but pod should be immediately terminated, to ensure other containers stop) | 
| `"Succeeded"` | `"Succeeded"` (but pod should be immediately terminated, to ensure other containers stop) |

File: neonvm/controllers/virtualmachine_controller.go (hunk 1 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:833]: Updated the function with a comment describing its semantics (eb8368ed5e3cbfb29162f4c9997eec3d39744769). Is that more consistent?

File: neonvm/controllers/virtualmachine_controller.go (hunk 2 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:872]: Changed the name to `runnerStatus` (1da6e055e8c6541325cbf59b5679608bb9dca11c), also to help differentiate from `runnerContainerStopped`. wdyt?

File: neonvm/controllers/virtualmachine_controller.go (hunk 6 above)
Comment (sharnoff) [reply neonvm/controllers/virtualmachine_controller.go:750]: Fair enough, reintroduced that explicit handling but kept the deduplication for the restarting logic (5c19ad6601ccedc2e2d1e32418b816dd1e005acd). wdyt?

Comment (shayanh) [reply neonvm/controllers/virtualmachine_controller.go:752]: Looks good

File: neonvm/controllers/virtualmachine_controller.go (hunk 2 above)
Comment (Omrigan) [reply neonvm/controllers/virtualmachine_controller.go:881]: Yes, makes sense

File: neonvm/controllers/virtualmachine_controller.go (hunk 1 above)
Comment (Omrigan) [reply neonvm/controllers/virtualmachine_controller.go:842]: Thanks, it's clear now.

File: neonvm/controllers/virtualmachine_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Omrigan) [reply neonvm/controllers/virtualmachine_controller.go:749]: That's a good table. I'd consider putting it to https://github.com/neondatabase/autoscaling/blob/main/neonvm/README.md

**Review by Omrigan — APPROVED** [review]

**Review by shayanh — APPROVED** [review]
